File: app/schemas/invoice.py
# =========================
# APP/SCHEMAS/INVOICE.PY
# =========================
"""
Pydantic v2 Schemas für Invoice.

Source of Truth für `/api/invoices/*`-Endpoints.
Gespiegelt vom SQLModel `Invoice` und `InvoiceItem` in `app/data.py:139-170`.

Status-Enum: DRAFT, OPEN, SENT, PAID, FINALIZED, CANCELLED.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


InvoiceStatus = Literal["DRAFT", "OPEN", "SENT", "PAID", "FINALIZED", "CANCELLED"]


class InvoiceItem(BaseModel):
    """Eine Rechnungsposition (description + quantity + unit_price)."""

    model_config = ConfigDict(extra="forbid")

    id: int | None = None
    description: str = Field(min_length=1, max_length=500)
    quantity: float = Field(gt=0, le=100000)
    unit_price: float = Field(ge=0, le=1_000_000_000)


class InvoiceDraft(BaseModel):
    """Input für POST /api/invoices (create) und /api/invoices/preview-pdf.

    Enthält alle vom Client kontrollierten Felder. Server-managed Felder
    (id, nr, status, total_brutto, updated_at, pdf_*) sind nicht enthalten.
    """

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    customer_id: int = Field(gt=0)
    title: str = Field(default="Rechnung", max_length=200)
    date: str = Field(default="", max_length=20)
    delivery_date: str = Field(default="", max_length=20)
    service_from: str = Field(default="", max_length=20)
    service_to: str = Field(default="", max_length=20)
    recipient_name: str = Field(default="", max_length=200)
    recipient_street: str = Field(default="", max_length=200)
    recipient_postal_code: str = Field(default="", max_length=20)
    recipient_city: str = Field(default="", max_length=100)
    vat_rate: float = Field(default=19.0, ge=0, le=100)
    ust_enabled: bool = Field(default=True)
    intro_text: str = Field(default="", max_length=2000)
    items: list[InvoiceItem] = Field(default_factory=list)
    notes: str = Field(default="", max_length=2000)
    status: InvoiceStatus = "OPEN"

    @field_validator("date", "delivery_date", "service_from", "service_to")
    @classmethod
    def _validate_iso_date(cls, v: str) -> str:
        if not v:
            return v
        from datetime import datetime
        try:
            datetime.fromisoformat(v)
        except ValueError:
            raise ValueError("Datum muss ISO-Format YYYY-MM-DD haben")
        return v


class InvoiceRead(BaseModel):
    """Output für GET /api/invoices[/...]. Vollständiger Record inkl. server-managed Felder."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    customer_id: int
    nr: str | None = None
    title: str = "Rechnung"
    date: str = ""
    delivery_date: str = ""
    recipient_name: str = ""
    recipient_street: str = ""
    recipient_postal_code: str = ""
    recipient_city: str = ""
    total_brutto: float
    status: InvoiceStatus = "DRAFT"
    revision_nr: int = 0
    updated_at: str = ""
    related_invoice_id: int | None = None
    items: list[InvoiceItem] = Field(default_factory=list)


class InvoiceStatusUpdate(BaseModel):
    """Input für PUT /api/invoices/{id}/status."""

    model_config = ConfigDict(extra="forbid")

    status: InvoiceStatus
    reason: str = Field(default="", max_length=500)


__all__ = [
    "InvoiceItem",
    "InvoiceDraft",
    "InvoiceRead",
    "InvoiceStatusUpdate",
    "InvoiceStatus",
]

File: frontend/src/lib/schemas/invoice.ts
import { z } from "zod";

/**
 * Zod-Mirror des Backend `app/schemas/invoice.py:InvoiceDraft`.
 * Source of Truth ist das Backend-Pydantic-Schema; dieses Zod-Schema
 * wird per `satisfies z.ZodType<InvoiceDraft>` validiert (siehe Backend-Tests).
 *
 * Wenn Backend oder Frontend divergieren, bricht entweder der Backend-Test
 * (`test_invoice_schemas.py`) oder der TypeScript-Build. So bleibt der Kontrakt
 * stabil ohne doppelte Definition.
 */

export const invoiceItemSchema = z.object({
  id: z.number().int().optional(),
  description: z.string().min(1, "Beschreibung erforderlich").max(500),
  quantity: z.number().positive("Menge muss > 0 sein").lte(100_000),
  unit_price: z.number().nonnegative("Preis darf nicht negativ sein").lte(1_000_000_000),
});

export type InvoiceItem = z.infer<typeof invoiceItemSchema>;

const isoDate = z
  .string()
  .regex(/^\d{4}-\d{2}-\d{2}$/, "Datum muss YYYY-MM-DD sein")
  .or(z.literal(""));

const optionalIsoDate = z
  .string()
  .regex(/^\d{4}-\d{2}-\d{2}$/)
  .or(z.literal(""));

export const invoiceStatusSchema = z.enum([
  "DRAFT",
  "OPEN",
  "SENT",
  "PAID",
  "FINALIZED",
  "CANCELLED",
]);

export type InvoiceStatus = z.infer<typeof invoiceStatusSchema>;

export const invoiceDraftSchema = z.object({
  customer_id: z.number().int().positive("Kunde erforderlich"),
  title: z.string().max(200).default("Rechnung"),
  subject: z.string().max(200).default(""),
  date: isoDate.default(""),
  delivery_date: optionalIsoDate.default(""),
  service_from: optionalIsoDate.default(""),
  service_to: optionalIsoDate.default(""),
  recipient_name: z.string().max(200).default(""),
  recipient_street: z.string().max(200).default(""),
  recipient_postal_code: z.string().max(20).default(""),
  recipient_city: z.string().max(100).default(""),
  vat_rate: z.number().nonnegative().lte(100).default(19.0),
  ust_enabled: z.boolean().default(true),
  intro_text: z.string().max(2000).default(""),
  items: z.array(invoiceItemSchema).default([]),
  notes: z.string().max(2000).default(""),
  status: invoiceStatusSchema.default("OPEN"),
});

export type InvoiceDraft = z.infer<typeof invoiceDraftSchema>;

export const emptyInvoiceDraft = (customerId: number): InvoiceDraft => ({
  customer_id: customerId,
  title: "Rechnung",
  subject: "",
  date: new Date().toISOString().slice(0, 10),
  delivery_date: "",
  service_from: "",
  service_to: "",
  recipient_name: "",
  recipient_street: "",
  recipient_postal_code: "",
  recipient_city: "",
  vat_rate: 19.0,
  ust_enabled: true,
  intro_text: "",
  items: [],
  notes: "",
  status: "OPEN",
});

/**
 * Read-side mirror of `app/schemas/invoice.py:InvoiceRead`.
 * Backend liefert: id, customer_id, nr, title, date, delivery_date,
 * recipient_*, total_brutto, status, revision_nr, updated_at,
 * related_invoice_id, items (optional, default []).
 */
export const invoiceReadSchema = z.object({
  id: z.number().int(),
  customer_id: z.number().int(),
  nr: z.string().nullable().optional(),
  title: z.string().default("Rechnung"),
  subject: z.string().default(""),
  date: z.string().default(""),
  delivery_date: z.string().default(""),
  recipient_name: z.string().default(""),
  recipient_street: z.string().default(""),
  recipient_postal_code: z.string().default(""),
  recipient_city: z.string().default(""),
  total_brutto: z.number().default(0),
  status: invoiceStatusSchema.default("OPEN"),
  revision_nr: z.number().int().default(0),
  updated_at: z.string().default(""),
  related_invoice_id: z.number().int().nullable().optional(),
  payment_link_url: z.string().default(""),
  payment_provider: z.string().default(""),
  items: z.array(invoiceItemSchema).default([]),
});

export type InvoiceRead = z.infer<typeof invoiceReadSchema>;
